src/esssans/normalization.py

```diff
@@ -64,6 +66,69 @@ def solid_angle_rectangular_approximation(
     )
 
 
+def solid_angle(
+    data: CalibratedMaskedData[RunType],
+    pixel_shape: DetectorPixelShape[RunType],
+    transform: SampleReferenceTransform[RunType],
+) -> SolidAngle[RunType]:
+    """
+    Solid angle for cylindrical pixels.
+
+    Note that the approximation is valid when the distance from sample
+    to pixel is much larger than the length or the radius of the pixels.
+
+    Parameters
+    ----------
+    data:
+        The DataArray that contains the positions of the detector pixels in coords.
+        The positions must be in the sample reference frame.
+
+    pixel_shape:
+        Contains the description of the detector pixel shape.
+
+    transform:
+        Transformation from the local reference frame of the detector
+        to the sample reference frame.
+
+    Returns
+    -------
+    :
+        The solid angle of the detector pixels, as viewed from the sample position.
+    """
+    face_1_center, face_1_edge, face_2_center = (
+        pixel_shape['vertices']['vertex', i] for i in range(3)
+    )
+    cylinder_axis = transform * face_2_center - transform * face_1_center
+    radius = sc.norm(face_1_center - face_1_edge)
+    length = sc.norm(cylinder_axis)
+
+    return SolidAngle[RunType](
+        approximate_solid_angle_for_cylinder_shaped_pixel_of_detector(
+            pixel_position=data.coords['position'],
+            cylinder_axis=cylinder_axis,
+            radius=radius,
+            length=length,
+        )
+    )
+
+
+def approximate_solid_angle_for_cylinder_shaped_pixel_of_detector(
+    pixel_position: sc.Variable,
+    cylinder_axis: sc.Variable,
+    radius: sc.Variable,
+    length: sc.Variable,
+):
+    """Computes solid angle of a detector pixel under the assumption that the
+    distance to the detector pixel from the sample is large compared to the radius
+    and the length of the piece of the detector cylinder that makes up the pixel.
+    """
+    norm_pp = sc.norm(pixel_position)
+    norm_ca = sc.norm(cylinder_axis)
+    cosalpha = sc.sqrt(1 - sc.dot(pixel_position, cylinder_axis) / (norm_pp * norm_ca))
+
+    return 2 * radius * length * cosalpha / norm_pp**2
```

Comment:
Not sure how much this matters here, but not that you could make this faster by avoiding a lot of large intermediate results, using, e.g., `+=` and `*=`. It generally get harder to read though.

There are also some easy gains to be made, e.g., by putting parenthesis (and/or reordering) in the last line, so ensure we first combine all the 0-D variables, to ensure we have to create fewer pixel-dependent intermediate results.

Reply:
I think it makes sense to group the 0-D variables, I'll fix that. But I don't think the performance improvement from using inplace operations will be worth the hit to readability in this case.